### DISCUSSION TITLE
New to hy, looking for another set of eyes on if statement

### DISCUSSION THREAD

**Discussion by bsturk** [hylang/hy#2551] [Q&A] [answered]
I'm writing a little client/server thing to learn Hy (loving it so far!), and there's an if statement that I don't understand why it is giving me an error, I hope someone can see what is going on, here's the code, the lines that are (attempting to be) bolded are the issue, and this is the error I get:

hostname[~/_drop_]% hy server.hy
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 288, in run_path
    code, fname = _get_code_from_file(run_name, path_name)
  File "/_drop_/server.hy", line 15
    (print "got CMD_HY_EVAL"))))
    ^
hy.errors.HySyntaxError: parse error for pattern macro 'if': got unexpected end of macro call, expected: form

```
(import socket)
(import json)

(defn handle-client [client-socket]
  (try
    (setv data (client-socket.recv 4096))
    (if (= (len data) 0)
      (print "No data received")
      (do
        (setv json-command (json.loads (data.decode "utf-8")))
        (print "Received JSON payload: " json-command)
        (setv cmd-type (get json-command "type"))
        (print "Command type: " cmd-type)
        **(if (= cmd-type "CMD_HY_EVAL")           <-----
          (print "got CMD_HY_EVAL"))))**
  (except [e Exception]
    (print "Exception: " (repr e))))
  (client-socket.close))

(defn start-server []
  (setv host "")
  (setv port 5000)
  (setv server-socket (socket.socket socket.AF_INET socket.SOCKS_STREAM))
  (server-socket.bind #(host port))
  (server-socket.listen 5)
  (print "Listening on port " port)
  (while True
    (try
       (setv [client-socket client_addr] (server-socket.accept))
       (handle-client client-socket)
    (except [k KeyboardInterrupt]
      (break))
    (except [e Exception]
      ; (print "timed out")))))
    ))))

(start-server)
```

**Comment by Kodiologist** [comment, marked as answer]
`if` needs a third argument.

**Reply by bsturk** [reply]
@Kodiologist thanks for that, so the else in an if is not optional good to know.

is there a way to express "nothing" in hy, equivalent to "pass" in python?

**Reply by Kodiologist** [reply]
We don't provide `pass` on the Hy side because it's just a syntactic wart in Python and not necessary in a free-form language; the Hy compiler implicitly generates `pass` when required by Python. For a conditional with no side-effect when the condition does not hold, one typically uses `when`, which is syntactic sugar for `if` with `None` for the else-form.

**Comment by bsturk** [comment]
that's perfect, thanks... any suggestions on how to handle an exception w/ a pass in Hy?  This is mostly being done for timeouts

  ; (print "timed out")))))

I would normally put a pass there as well, it seems putting nothing there works, but I wanted to make sure I'm doing it correctly.  In this case, the exception should be a no-op.  I am ignoring socket timeout exceptions, if I don't have the handler the server stops.

If I'm reading your comment correctly though, a "pass" is being implicitly generated there?

**Reply by Kodiologist** [reply]
You posted this as an answer to your own question. Perhaps you meant to reply to my comment or to create a new discussion.
